// genrate/dist/templates/lint/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.lintstaged = exports.commitlint = exports.prettier = exports.eslint = exports.czconf = exports.husky = exports.cz = void 0;
var czrc_1 = require("./czrc");
Object.defineProperty(exports, "cz", { enumerable: true, get: function () { return __importDefault(czrc_1).default; } });
var husky_1 = require("./husky");
Object.defineProperty(exports, "husky", { enumerable: true, get: function () { return __importDefault(husky_1).default; } });
var czconfig_1 = require("./czconfig");
Object.defineProperty(exports, "czconf", { enumerable: true, get: function () { return __importDefault(czconfig_1).default; } });
var eslintrc_1 = require("./eslintrc");
Object.defineProperty(exports, "eslint", { enumerable: true, get: function () { return __importDefault(eslintrc_1).default; } });
var prettierrc_1 = require("./prettierrc");
Object.defineProperty(exports, "prettier", { enumerable: true, get: function () { return __importDefault(prettierrc_1).default; } });
var commitlintrc_1 = require("./commitlintrc");
Object.defineProperty(exports, "commitlint", { enumerable: true, get: function () { return __importDefault(commitlintrc_1).default; } });
var lintstagedrc_1 = require("./lintstagedrc");
Object.defineProperty(exports, "lintstaged", { enumerable: true, get: function () { return __importDefault(lintstagedrc_1).default; } });
