// genrate/dist/configs/parser/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.bundlerConfig = exports.ignoreConfig = exports.vsCodeConfig = exports.lintConfig = exports.testConfig = exports.e2eConfig = exports.pkgConfig = exports.tsConfig = void 0;
var ts_1 = require("./ts");
Object.defineProperty(exports, "tsConfig", { enumerable: true, get: function () { return __importDefault(ts_1).default; } });
var pkg_1 = require("./pkg");
Object.defineProperty(exports, "pkgConfig", { enumerable: true, get: function () { return __importDefault(pkg_1).default; } });
var e2e_1 = require("./e2e");
Object.defineProperty(exports, "e2eConfig", { enumerable: true, get: function () { return __importDefault(e2e_1).default; } });
var test_1 = require("./test");
Object.defineProperty(exports, "testConfig", { enumerable: true, get: function () { return __importDefault(test_1).default; } });
var lint_1 = require("./lint");
Object.defineProperty(exports, "lintConfig", { enumerable: true, get: function () { return __importDefault(lint_1).default; } });
var vscode_1 = require("./vscode");
Object.defineProperty(exports, "vsCodeConfig", { enumerable: true, get: function () { return __importDefault(vscode_1).default; } });
var ignore_1 = require("./ignore");
Object.defineProperty(exports, "ignoreConfig", { enumerable: true, get: function () { return __importDefault(ignore_1).default; } });
var bundler_1 = require("./bundler");
Object.defineProperty(exports, "bundlerConfig", { enumerable: true, get: function () { return __importDefault(bundler_1).default; } });
