// genrate/dist/templates/ts/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const lodash_1 = require("lodash");
const utils_1 = require("../../utils");
const defaultOptions = {
    compilerOptions: {
        strict: true,
        lib: ['ESNext'],
        target: 'ESNext',
        module: 'ESNext',
        skipLibCheck: true,
        importHelpers: true,
        removeComments: true,
        esModuleInterop: true,
        moduleResolution: 'node',
        resolveJsonModule: true,
        useDefineForClassFields: true,
        forceConsistentCasingInFileNames: true
    },
    include: ['src/**/*.ts']
};
exports.default = {
    node: (0, utils_1.stringify)(defaultOptions),
    web: (0, utils_1.stringify)((0, lodash_1.merge)({}, defaultOptions, { compilerOptions: { lib: ['ESNext', 'DOM'] } })),
    vue: (fvs) => (0, utils_1.stringify)((0, lodash_1.merge)({}, defaultOptions, {
        include: fvs === '2.x'
            ? ['src/**/*', 'src/**/*.vue', 'test/**/*']
            : ['src/**/*', 'env.d.ts', 'src/**/*.vue', 'vite.config.ts', 'test/**/*'],
        compilerOptions: {
            jsx: 'preserve',
            isolatedModules: fvs === '2.x' ? undefined : true,
            experimentalDecorators: fvs === '2.x' ? true : undefined,
            preserveValueImports: true,
            importsNotUsedAsValues: 'error',
            lib: ['ESNext', 'DOM', 'DOM.Iterable'],
            baseUrl: '.',
            types: ['node'],
            paths: { '@/*': ['./src/*'] }
        }
    })),
    lib: {
        node: (0, utils_1.stringify)((0, lodash_1.merge)({}, defaultOptions, { compilerOptions: { declaration: true, declarationDir: 'types' } })),
        web: (0, utils_1.stringify)((0, lodash_1.merge)({}, defaultOptions, { compilerOptions: { declaration: true, declarationDir: 'types', lib: ['ESNext', 'DOM'] } })),
        vue: (fvs) => (0, utils_1.stringify)((0, lodash_1.merge)({}, defaultOptions, {
            include: fvs === '2.x'
                ? ['src/**/*', 'src/**/*.vue', 'test/**/*']
                : ['src/**/*', 'env.d.ts', 'src/**/*.vue', 'vite.config.ts', 'test/**/*'],
            compilerOptions: {
                jsx: 'preserve',
                isolatedModules: fvs === '2.x' ? undefined : true,
                experimentalDecorators: fvs === '2.x' ? true : undefined,
                preserveValueImports: true,
                importsNotUsedAsValues: 'error',
                lib: ['ES2016', 'DOM', 'DOM.Iterable'],
                baseUrl: '.',
                types: ['node'],
                paths: { '@/*': ['./src/*'] }
            }
        }))
    }
};
